server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java

```diff
@@ -347,4 +357,113 @@ public String toString() {
             }
         }
     }
+
+    private class ConnectionHistory {
+        record NodeConnectionHistory(String ephemeralId, long disconnectTime, Exception disconnectCause) {}
+
+        /**
+         * Holds the DiscoveryNode nodeId to connection history record.
+         *
+         * Entries for each node are reserved during NodeConnectionsService.connectToNodes, by placing a (nodeId, dummy) entry
+         * for each node in the cluster. On node disconnect, this entry is updated with its NodeConnectionHistory. On node
+         * connect, this entry is reset to the dummy value. On NodeConnectionsService.disconnectFromNodesExcept, node entries
+         * are removed.
+         *
+         * Each node in the cluster always has a nodeHistory entry that is either the dummy value or a connection history record. This
+         * allows node disconnect callbacks to discard their entry if the disconnect occurred because of a change in cluster state.
+         */
+        private final NodeConnectionHistory dummy = new NodeConnectionHistory("", 0, null);
```

Comment:
Can be `static` I think, it's a global constant. We tend to name global constants in `SHOUTY_SNAKE_CASE` reflecting their meaning, so here I'd suggest `CONNECTED` or `CONNECTED_MARKER` or something like that. This way you get to say `nodeConnectionHistory != CONNECTED_MARKER` below which makes it clearer to the reader what this predicate means.

Reply:
nit: also looks like the javadoc is for the `nodeHistory` field